### DISCUSSION TITLE
Mapping onto an existing object with children

### DISCUSSION THREAD

**Discussion by jpbertles** [riok/mapperly#1310] [Q&A]
First off, I just want to say that this is an awesome library. Really great work on this!

I have a situation where I have a Customer object that has an Address. I would like to map onto an existing customer and for the customer specific properties everything is working great, but the auto generated code always creates a new address object, and doesn't map onto the existing one. Is there anything I can do to ensure the existing address object is mapped onto rather than creating a new one?

```csharp
public class Customer
{
    public Guid CustomerId { get; set; }
    public string Name { get; set; }
    ...
    public Address Address { get; set; }
}

public class Address
{
    public Guid AddressId { get; set; }
    public string Street1 { get; set; }
    ....
}

public class CustomerDto
{
    public string Name { get; set; }
    ...
    public AddressDto Address { get; set; }
}

public class AddressDto
{
    public string Street1 { get; set; }
    ....
}

[Mapper]
public static partial CustomerMapper
{
    public static partial void MapUpdateOntoDbModel(CustomerDto dto, Customer customer);
}


/// This is the generated code
public static partial class CustomerMapper
{
    [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.5.1.0")]
    public static partial void MapUpdateOntoDbModel(global::CustomerDto dto, global:Customer customer)
    {
        if (dto.Address != null)
        {
            customer.Address = MapToAddress1(dto.Address);
        }
        else
        {
           customer.Address = null;
        }
        ....
    }
}
```

**Comment by latonz** [comment]
Created #1311

**Comment by ste4net** [comment]
I think there's something I'm missing because this request doesn't seem to have the audience I expected, so I don't understand how others handle enumerable properties.

All my `List<T>` properties need to always be managed with updates. If the source has new elements, I add them; if the target has extra elements compared to the source, I remove them, and for the rest, I update them. The comparison is based on the unique key of the element.

I don't see any other way this could be managed.
@jpbertles  Are you using Mapperly? If so, how do you handle this situation?

**Reply by latonz** [reply]
How does your manually written code look like to achieve this? This sound quite complex and I'm not sure if a general object to object mapper is the right place to do complex merging operations. May also be related to https://github.com/riok/mapperly/issues/665.

**Reply by ste4net** [reply]
Yes it is related to #665  I would use an approach like the one I’ve written below (done quickly and not optimized, just as an example).
But I imagine I could solve the list mapping by implementing a generic method on the base class (it's related to post 234).
However, the main problem is that the mapping between existing objects in the case of lists always creates a new list, or it expects custom methods to create a new list instead of a mapping method that accepts the existing list.

```csharp
        public void MapExisting(Company src, CompanyViewModel dst)
         {

            //...
            //if (dst.Addresses == null) dst.Addresses = new List<AddressVM>(); //Redoundat if initialized in constructor
            MapToList(src.Addresses, dst.Addresses, (x, y) => x.Id == y.Id);
        }
        public void MapListElemtxy<T1, T2>(T1 src, T2 dst)
        {
            //...
        }
        public void MapToList<T, T2>(List<T> Src, List<T2> Dst, Func<T, T2, bool> predicate) where T : class, IDomainEntity where T2 : class, IDomainEntity, new()
        {
            // Remove all items from Dst not in Src
            Dst.RemoveAll(x => !Src.Any(y => y.Id == x.Id));
            foreach (var item in Src)
            {
                T2 AddOrUpd = Dst.FirstOrDefault(d => d.Id == item.Id, new T2());
                MapListElemtxy(item, AddOrUpd);
                if (AddOrUpd.Id == 0) Dst.Add(AddOrUpd);
            }
        }
```

Now mapperly produce this ...

```
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "3.6.0.0")]
        public partial void DomainToVm(global::BizWork.LohiCore.Domain.MBase.Company domain, global::LohiCore.Server.ViewModels.CompanyViewModel vm)
        {
            if (domain.Addresses != null)
            {
                vm.Addresses = MapToListOfAddress(domain.Addresses);
            }
            else
            {
                vm.Addresses = null;
            }
```
Instead when we are mapping between existng object it should be something like

```
 MapToListOfExistingAddress(domain.Addresses,vm.Addresses);
Or
MapToListOfExistngItems<T,T2>(domain.Addresses,vm.Addresses);
```
Just to have the possibility of implementing the list update if it is considered too specific a case to be implemented in the mapper

**Reply by latonz** [reply]
This is tracked in https://github.com/riok/mapperly/issues/451
